File: 04-advent-of-code-2019/day01.py
import pytest


@pytest.mark.parametrize("mass, fuel", [
    (12, 2),
    (14, 2),
    (1969, 654),
    (100756, 33583),
    (2, 0),
])
def test_required_fuel(mass, fuel):
    assert required_fuel(mass) == fuel


def required_fuel(mass):
    return max(0, mass // 3 - 2)


def part1(values):
    return sum(required_fuel(mass) for mass in values)


@pytest.mark.parametrize("mass, fuel", [
    (14, 2),
    (1969, 966),
    (100756, 50346),
])
def test_required_fuel_taking_fuel_mass_into_account(mass, fuel):
    assert required_fuel_taking_fuel_mass_into_account(mass) == fuel


def required_fuel_taking_fuel_mass_into_account(mass):
    extra_mass = []
    while True:
        mass = required_fuel(mass)
        if mass == 0:
            break
        extra_mass.append(mass)
    return sum(extra_mass)


def part2(values):
    return sum(required_fuel_taking_fuel_mass_into_account(mass) for mass in values)


if __name__ == '__main__':
    with open("day01.txt") as file_:
        values = [int(line) for line in file_]
    print("Part 1:", part1(values))
    print("Part 2:", part2(values))

File: 04-advent-of-code-2019/day01.pl
%% SWI Prolog version
%%
%% Run tests:
%%   $ swipl -q -t run_tests day01.pl
%%
%% Find solution for part 1:
%%   $ swipl -q -t part1 day01.pl

required_fuel(Mass, Fuel) :- Fuel is max(0, Mass // 3 - 2).

:- begin_tests(tests).
test(required_fuel, true(F == 2)) :- required_fuel(12, F).
test(required_fuel, true(F == 2)) :- required_fuel(14, F).
test(required_fuel, true(F == 654)) :- required_fuel(1969, F).
test(required_fuel, true(F == 33583)) :- required_fuel(100756, F).
test(required_fuel, true(F == 0)) :- required_fuel(2, F).
:- end_tests(tests).

read_input(Filename, Values) :-
    read_file_to_string(Filename, String, []),
    split_string(String, "\n", "", Lines),
    maplist(number_string, Values, Lines).

total_required_fuel(MassList, TotalFuel) :-
    maplist(required_fuel, MassList, FuelList),
    sum_list(FuelList, TotalFuel).

part1 :-
    read_input("day01.txt", MassList),
    forall(total_required_fuel(MassList, TotalFuel), writeln(TotalFuel)).
